### ISSUE TITLE
fancy indexing: error with empty result

### ISSUE THREAD

**Issue by brentp** [mratsim/Arraymancer#445]
```Nim
var r = @[1'f32, 1'f32].toTensor()
echo r[r <. 0]
```
gives: 
```
Error: unhandled exception: index out of bounds, the container is empty [IndexDefect]
```

**Comment by Vindaar** [comment]
Hm, I tried to fix this right now, but I'm running into problems on how to handle empty tensors in arraymancer. 

So the underlying problem of this issue can be fixed easily by returning an empty tensor if the input has no true values / is an empty tensor / an empty seq. But then the resulting empty tensor has some weird behavior where it will run into all sorts of issues, because many of the low level operations expect the data storage to have elements in them. 

I'm not sure if the solution is to check in `p_accessors.nim` for size of the tensors too? I don't have a good overview on how those templates are used. Some input @mratsim would be appreciated.

**Comment by Vindaar** [comment]
Just saw this issue: https://github.com/mratsim/Arraymancer/issues/210

**Comment by mratsim** [comment]
Related: https://github.com/pytorch/pytorch/issues/5014

**Comment by mratsim** [comment]
And seems like it's messy on Intel side as well: https://github.com/oneapi-src/oneDNN/issues/217

**Comment by mratsim** [comment]
It seems like Numpy also IndexError on empty tensors: https://stackoverflow.com/questions/53747948/what-does-a-numpy-shape-starting-with-zero-mean

I'll try to copy numpy behavior on the basic reshaping and arithmetic operations implemented. Hopefully they make sense.

**Comment by mratsim** [comment]
Now if someone can explain why this reliably returns 0.007812501848093234 ...

```
import numpy as np

def empty_add():
    empty = np.ndarray([])

    print(empty)

if __name__ == "__main__":
    empty_add()
```

**Comment by mratsim** [comment]
Closed by #448
